functionary/vllm_monkey_patch/async_llm_engine.py

```diff
@@ -0,0 +1,724 @@
+                    ]
+
+                new_curr_tokens = curr_text + self.tokenizer.decode([sampled_token_ind])
+
+                options_mask = []
+                for option in options:
+                    if option not in wellformed_params and option.startswith(
+                        new_curr_tokens
+                    ):
+                        options_mask.append(True)
+                    else:
+                        options_mask.append(False)
+
+                # Same logic as function name, except that we check whether the token
+                # is a stopping token for parameter name generation.
+                if (
+                    sampled_token_ind
+                    == prompt_template.get_stopping_token(stage="parameter")
+                    or any(options_mask)
+                ) and sampled_token.strip(" ") != "":
+                    return sampled_token_ind, sampled_token
+
+    async def step_async(self) -> List[RequestOutput]:
+        """Performs one decoding iteration and returns newly generated results.
+        The workers are ran asynchronously if possible.
+
+        This function performs one decoding iteration of the engine. It first
+        schedules the sequences to be executed in the next iteration and the
+        token blocks to be swapped in/out/copy. Then, it executes the model
+        and updates the scheduler with the model outputs. Finally, it decodes
+        the sequences and returns the newly generated results.
+        """
+        seq_group_metadata_list, scheduler_outputs, ignored = self._schedule()
+        if scheduler_outputs.is_empty():
+            return ignored
+
+        # Execute the model.
+        output = await self._run_workers_async(
+            "execute_model",
+            seq_group_metadata_list=seq_group_metadata_list,
+            blocks_to_swap_in=scheduler_outputs.blocks_to_swap_in,
+            blocks_to_swap_out=scheduler_outputs.blocks_to_swap_out,
+            blocks_to_copy=scheduler_outputs.blocks_to_copy,
+        )
+
+        # Loop through all the output in the batch
+        for i in range(len(output)):
+            # Get all the required variables for grammar sampling
+            request_id = seq_group_metadata_list[i].request_id
+
+            model_sampled_token_id = output[i].samples[-1].output_token
+            delta_token_id_by_logprobs = list(output[i].samples[-1].logprobs.keys())
+
+            # Perform grammar sampling if needed
+            grammar_sampled_token_id, grammar_sampled_token = self.sample(
+                request_id=request_id,
+                delta_token_ids=delta_token_id_by_logprobs,
+                model_sampled_token_id=model_sampled_token_id,
+            )
+
+            # Update the output token to vllm with the newly sampled one
+            output[i].samples[-1].output_token = grammar_sampled_token_id
+
+            # Update gen_state
+            self.update_gen_state(
```

Comment:
Can we also include update_gen_state inside def sample, I think it would be more convenient, no need to duplicate the this chunk of code:
```
if gen_state["stage"] in ["pre-function", "function"]:
            options = [
                tool_or_func["name"]
                for tool_or_func in self.tools_or_functions[request_id]
            ]
else:
    func_name = gen_state["func_name"]
    for tool_or_func in self.tools_or_functions[request_id]:
        if tool_or_func["name"] == func_name:
            options = list(tool_or_func["parameters"]["properties"].keys())
```

Reply:
I will put this chunk into `step_async` before calling `prompt_template.step_async`.